This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [10]:
# get api key from your .env file
import os
from dotenv import load_dotenv  # if missing this module, simply run `pip install python-dotenv`

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')



Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [11]:
# First, import the relevant modules
import requests
import json
import sys

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [12]:
API_KEY = '3cWCRetTpfJvT3ofzBCv'
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
one_day_url = f"https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X.json"
params = {'start_date': '2018-11-30',
          'end_date': '2018-11-30',
          'api_key': API_KEY
          }

r = requests.get(one_day_url, params)
r.text

'{"dataset":{"id":10095370,"dataset_code":"AFX_X","database_code":"FSE","name":"Carl Zeiss Meditec (AFX_X)","description":"Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.\\u003cbr\\u003e\\u003cbr\\u003eTrading System: Xetra\\u003cbr\\u003e\\u003cbr\\u003eISIN: DE0005313704","refreshed_at":"2020-12-01T14:48:09.907Z","newest_available_date":"2020-12-01","oldest_available_date":"2000-06-07","column_names":["Date","Open","High","Low","Close","Change","Traded Volume","Turnover","Last Price of the Day","Daily Traded Units","Daily Turnover"],"frequency":"daily","type":"Time Series","premium":false,"limit":null,"transform":null,"column_index":null,"start_date":"2018-11-30","end_date":"2018-11-30","data":[["2018-11-30",null,62.3,60.95,61.0,null,142218.0,8716082.0,null,null,null]],"collapse":null,"order":null,"database_id":6129}}'

{"dataset":{"id":10095370,"dataset_code":"AFX_X","database_code":"FSE","name":"Carl Zeiss Meditec (AFX_X)","description":"Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.\u003cbr\u003e\u003cbr\u003eTrading System: Xetra\u003cbr\u003e\u003cbr\u003eISIN: DE0005313704","refreshed_at":"2020-12-01T14:48:09.907Z","newest_available_date":"2020-12-01","oldest_available_date":"2000-06-07","column_names":["Date","Open","High","Low","Close","Change","Traded Volume","Turnover","Last Price of the Day","Daily Traded Units","Daily Turnover"],"frequency":"daily","type":"Time Series","premium":false,"limit":null,"transform":null,"column_index":null,"start_date":"2018-11-30","end_date":"2018-11-30","data":[["2018-11-30",null,62.3,60.95,61.0,null,142218.0,8716082.0,null,null,null]],"collapse":null,"order":null,"database_id":6129}}

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

Task 1 -  Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).

In [13]:
#Task 1 -  Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
url_afx_2017 = f"https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X.json"

params = {'start_date': '2017-01-01',
          'end_date': '2017-12-31',
          'api_key': API_KEY
          }
request = requests.get(url_afx_2017, params)
# filename = "afx.json"
# with open(filename,"w") as outfile:
#     outfile.write(request.text)

In [14]:
#2. Convert the returned JSON object into a Python dictionary.

# with open(filename, "r") as openfile_json:
#     afx_2017_json = json.load(openfile_json)
afx_2017_json = json.loads(request.text)
    

In [15]:
def median(list):
    list.sort()
    items = len(list)
    middle = items//2
    
    if items % 2 == 1:
        return list[middle]
    else:
        return (list[middle-1] + list[middle])/2
       
    

In [16]:
#test median
# print(median([1,2,3]))
# print(median([1,2]))
# print(median([1,2,3,4]))
# print(median([1,2,3,4,5]))
# print(median([1,2,3,4,5,6]))
# print(median([1,2,3,4,5,6,7]))

In [19]:
#3. Calculate what the highest and lowest opening prices were for the stock in this period.
data = afx_2017_json["dataset"]["data"]

#3. Calculate what the highest and lowest opening prices were for the stock in this period.
highest_opening_price = 0
lowest_opening_price = sys.maxsize

#4. What was the largest change in any one day (based on High and Low price)?
largest_difference_between_high_and_low = 0

#5. What was the largest change between any two days (based on Closing Price)?
last_close = 0
max_last_close = 0

#6. What was the average daily trading volume during this year?
total_traded_volume = 0

#7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)
volume = []

for date, open, high, low, close, change, traded_volume, turn_over, last_price_of_the_day, daily_traded_units, daily_turnover in data:

    #3
    if open:
        highest_opening_price = max(open,highest_opening_price)
        lowest_opening_price = min(open,lowest_opening_price)
    
    #4 
    if high and low:
        largest_difference_between_high_and_low = max(
            high-low, largest_difference_between_high_and_low)
        
    #5
    if close and last_close:
        max_last_close = max(abs(last_close-close),max_last_close)        
    
    last_close = close
    
    #6
    total_traded_volume += traded_volume
    
    #7
    volume.append(traded_volume)

#6
average_traded_volume = total_traded_volume / len(data)

#7
median_volume = median(volume)
        
padding = 10
print(f"Highest opening prices : {highest_opening_price: >{padding}.2f}")
print(f"Lowest opening price   : {lowest_opening_price: >{padding}.2f}")
print(f"Largest difference     : {largest_difference_between_high_and_low: >{padding}.2f}")
print(f"Largest change         : {max_last_close: >{padding}.2f}")
print(f"Average traded volume  : {average_traded_volume: >{padding},.2f}")
print(f"Median volume          : {median_volume: >{padding},.2f}")


Highest opening prices :      53.11
Lowest opening price   :      34.00
Largest difference     :       2.81
Largest change         :       2.56
Average traded volume  :  89,124.34
Median volume          :  76,286.00
